subcommand/tls-init/command_test.go

```diff
@@ -0,0 +1,632 @@
+	newServerKey := serverCertSecret.Data[corev1.TLSPrivateKeyKey]
+
+	certBlock, _ := pem.Decode(newServerCert)
+	certificate, err := x509.ParseCertificate(certBlock.Bytes)
+	require.NoError(t, err)
+	require.False(t, certificate.IsCA)
+	require.Equal(t, []string{"server.dc1.consul", "localhost"}, certificate.DNSNames)
+	require.Equal(t, []net.IP{net.ParseIP("127.0.0.1").To4()}, certificate.IPAddresses)
+
+	keyBlock, _ := pem.Decode(newServerKey)
+	privateKey, err := x509.ParseECPrivateKey(keyBlock.Bytes)
+	require.NoError(t, err)
+	require.Equal(t, &privateKey.PublicKey, certificate.PublicKey)
+}
+
+func TestRun_UpdatesServerCertificatesWithExistingCertsAsSecrets(t *testing.T) {
+	ui := cli.NewMockUi()
+	cmd := Command{UI: ui}
+	k8s := fake.NewSimpleClientset()
+	cmd.clientset = k8s
+
+	_, err := k8s.CoreV1().Secrets("default").Create(context.Background(), &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "consul-ca-cert",
+			Namespace: "default",
+		},
+		Data: map[string][]byte{
+			corev1.TLSCertKey: []byte(caCert),
+		},
+		Type: corev1.SecretTypeOpaque,
+	}, metav1.CreateOptions{})
+	require.NoError(t, err)
+
+	_, err = k8s.CoreV1().Secrets("default").Create(context.Background(), &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "consul-ca-key",
+			Namespace: "default",
+		},
+		Data: map[string][]byte{
+			corev1.TLSPrivateKeyKey: []byte(caKey),
+		},
+		Type: corev1.SecretTypeOpaque,
+	}, metav1.CreateOptions{})
+	require.NoError(t, err)
+
+	_, err = k8s.CoreV1().Secrets("default").Create(context.Background(), &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "consul-server-cert",
+			Namespace: "default",
+		},
+		Data: map[string][]byte{
+			corev1.TLSCertKey:       []byte(serverCert),
+			corev1.TLSPrivateKeyKey: []byte(serverKey),
+		},
+		Type: corev1.SecretTypeTLS,
+	}, metav1.CreateOptions{})
+	require.NoError(t, err)
+
+	flags := []string{"-name-prefix", "consul", "-additional-dnsname", "test.dns.name"}
+	exitCode := cmd.Run(flags)
+	require.Equal(t, 0, exitCode)
+
+	serverCertSecret, err := k8s.CoreV1().Secrets("default").Get(context.Background(), "consul-server-cert", metav1.GetOptions{})
+	require.NoError(t, err)
+	newServerCert := serverCertSecret.Data[corev1.TLSCertKey]
+	newServerKey := serverCertSecret.Data[corev1.TLSPrivateKeyKey]
+
+	require.NotEqual(t, []byte(serverCert), newServerCert)
+	certBlock, _ := pem.Decode(newServerCert)
+	certificate, err := x509.ParseCertificate(certBlock.Bytes)
+	require.NoError(t, err)
+	require.False(t, certificate.IsCA)
+	require.Equal(t, []string{"test.dns.name", "server.dc1.consul", "localhost"}, certificate.DNSNames)
+
+	require.NotEqual(t, []byte(serverKey), newServerKey)
+	keyBlock, _ := pem.Decode(newServerKey)
+	privateKey, err := x509.ParseECPrivateKey(keyBlock.Bytes)
+	require.NoError(t, err)
+	require.Equal(t, &privateKey.PublicKey, certificate.PublicKey)
+}
+
+func TestRun_DoesNotRecreateExistingCACertSecrets(t *testing.T) {
```

Comment:
I thought that since we pre-create the CA secret, then run the command and then make assertions that the certificate's public key is equal to the CA's public key, that's essentially the same. That assertion would fail if a new CA was generated instead of using the existing CA in the secret. I could be missing something though!

Reply:
Aah. That wasn't an assertion here but I could add that one. The assertion made was to compare the public key of the certificate with the public key for the private key generated. But I think this is a good assertion to add that should address all of these.

Reply:
Updated the test to verify that the certificate is issued by the CA.